File: lib/models/tooth.dart
import 'dart:convert';

import 'package:equatable/equatable.dart';
import 'package:get/get.dart';

class Tooth extends Equatable {
  final String id;
  final ToothType type;
  final ToothCondition condition;
  final String imagePath;

  const Tooth({
    required this.id,
    required this.type,
    required this.condition,
    required this.imagePath,
  });

  String get icon {
    switch (type) {
      case ToothType.SERI_1:
      case ToothType.SERI_2:
      case ToothType.TARING:
        return getClosedIcon();
      default:
        return getOpenIcon();
    }
  }

  String getClosedIcon() {
    switch (condition) {
      case ToothCondition.NORMAL:
        return "assets/normal1.png";
      case ToothCondition.IMPAKSI:
        return "assets/impaksi1.png";
      case ToothCondition.KARIES:
        return "assets/karies1.png";
      case ToothCondition.SISA_AKAR:
        return "assets/sisa_akar1.png";
      case ToothCondition.TUMPATAN:
        return "assets/tumpatan1.png";
    }
  }

  String getOpenIcon() {
    switch (condition) {
      case ToothCondition.NORMAL:
        return "assets/normal.png";
      case ToothCondition.IMPAKSI:
        return "assets/impaksi.png";
      case ToothCondition.KARIES:
        return "assets/karies.png";
      case ToothCondition.SISA_AKAR:
        return "assets/sisa_akar.png";
      case ToothCondition.TUMPATAN:
        return "assets/tumpatan.png";
    }
  }

  String getTitle() => "$id - ${type.name.replaceFirst("_", " ").toLowerCase().capitalizeFirst}";

  @override
  List<Object?> get props => [id, type, condition, imagePath];
}

enum ToothType {
  SERI_1,
  SERI_2,
  TARING,
  PREMOLAR_1,
  PREMOLAR_2,
  MOLAR_1,
  MOLAR_2,
  MOLAR_3
}

enum ToothCondition { NORMAL, KARIES, TUMPATAN, SISA_AKAR, IMPAKSI }

enum ToothQuadrant { QUADRANT_I, QUADRANT_II, QUADRANT_III, QUADRANT_IV }

extension ToothQuadranExt on ToothQuadrant {
  List<int> get idList => switch (this) {
        ToothQuadrant.QUADRANT_I => List.generate(8, (index) => index + 11),
        ToothQuadrant.QUADRANT_II => List.generate(8, (index) => index + 21),
        ToothQuadrant.QUADRANT_III => List.generate(8, (index) => index + 41),
        ToothQuadrant.QUADRANT_IV => List.generate(8, (index) => index + 31),
      };

  bool get isReverse =>
      [ToothQuadrant.QUADRANT_I, ToothQuadrant.QUADRANT_III].contains(this);

  String get title => name.toLowerCase().split("_").join(" ").capitalize ?? "";

  String get icon => switch (this) {
        ToothQuadrant.QUADRANT_I => "assets/kuadran_1.png",
        ToothQuadrant.QUADRANT_II => "assets/kuadran_2.png",
        ToothQuadrant.QUADRANT_III => "assets/kuadran_3.png",
        ToothQuadrant.QUADRANT_IV => "assets/kuadran_4.png",
      };
  
  int get intRepresentative => switch(this) {
    ToothQuadrant.QUADRANT_I => 1,
    ToothQuadrant.QUADRANT_II => 2,
    ToothQuadrant.QUADRANT_III => 3,
    ToothQuadrant.QUADRANT_IV => 4
  };
}

extension ToothTypeIdExt on int {
  ToothType get toothType => ToothType.values[this % 10 - 1];
}

extension StringExt on String {
  List<Tooth> parseToToothList() {
    final List<dynamic> jsonList = jsonDecode(this);
    
    final List<Tooth> toothList = jsonList.map((jsonObject) {
      return Tooth(
        id: jsonObject['id'],
        type: _parseToothType(jsonObject['type']),
        condition: _parseToothCondition(jsonObject['condition']),
        imagePath: jsonObject['imagePath'],
      );
    }).toList();

    return toothList;
  }

  ToothType _parseToothType(String typeString) {
    return ToothType.values.firstWhere(
      (type) => type.toString().split('.').last == typeString,
      orElse: () => ToothType.SERI_1,
    );
  }

  ToothCondition _parseToothCondition(String conditionString) {
    return ToothCondition.values.firstWhere(
      (condition) => condition.toString().split('.').last == conditionString,
      orElse: () => ToothCondition.NORMAL,
    );
  }
}

File: android/app/src/main/java/com/example/odontogram/domain/entity/Tooth.kt
package com.example.odontogram.domain.entity

import com.example.odontogram.R
import com.example.odontogram.domain.entity.ToothCondition.IMPAKSI
import com.example.odontogram.domain.entity.ToothCondition.KARIES
import com.example.odontogram.domain.entity.ToothCondition.NORMAL
import com.example.odontogram.domain.entity.ToothCondition.SISA_AKAR
import com.example.odontogram.domain.entity.ToothCondition.TUMPATAN
import com.example.odontogram.domain.entity.ToothType.SERI_1
import com.example.odontogram.domain.entity.ToothType.SERI_2
import com.example.odontogram.domain.entity.ToothType.TARING
import kotlinx.serialization.Serializable

@Serializable
data class Tooth(
    val id: String,
    val type: ToothType,
    val condition: ToothCondition,
    val imagePath: String
) {
    val icon =
        when (type) {
            SERI_1, SERI_2, TARING -> getClosedIcon()
            else -> getOpenIcon()
        }

    private fun getClosedIcon() = when (condition) {
        NORMAL -> R.drawable.normal1
        IMPAKSI -> R.drawable.impaksi1
        KARIES -> R.drawable.karies1
        SISA_AKAR -> R.drawable.sisa_akar1
        TUMPATAN -> R.drawable.tumpatan1
    }


    private fun getOpenIcon() = when (condition) {
        NORMAL -> R.drawable.normal
        IMPAKSI -> R.drawable.impaksi
        KARIES -> R.drawable.karies
        SISA_AKAR -> R.drawable.sisa_akar
        TUMPATAN -> R.drawable.tumpatan
    }
}

@Serializable
enum class ToothType(
    val q1: Int,
    val q2: Int,
    val q3: Int,
    val q4: Int
) {
    SERI_1(11, 21, 41, 31),
    SERI_2(12, 22, 42, 32),
    TARING(13, 23, 43, 33),
    PREMOLAR_1(14, 24, 44, 34),
    PREMOLAR_2(15, 25, 45, 35),
    MOLAR_1(16, 26, 46, 36),
    MOLAR_2(17, 27, 47, 37),
    MOLAR_3(18, 28, 48, 38)
}

fun ToothType.getId(quadrant: ToothQuadrant) = when (quadrant) {
    ToothQuadrant.QUADRANT_I -> q1
    ToothQuadrant.QUADRANT_II -> q2
    ToothQuadrant.QUADRANT_III -> q3
    ToothQuadrant.QUADRANT_IV -> q4
}

@Serializable
enum class ToothCondition {
    NORMAL,
    KARIES,
    TUMPATAN,
    SISA_AKAR,
    IMPAKSI
}

enum class ToothQuadrant {
    QUADRANT_I,
    QUADRANT_II,
    QUADRANT_III,
    QUADRANT_IV
}

fun ToothQuadrant.getIdList() = when (this) {
    ToothQuadrant.QUADRANT_I -> (11..18).toList()
    ToothQuadrant.QUADRANT_II -> (21..28).toList()
    ToothQuadrant.QUADRANT_III -> (41..48).toList()
    ToothQuadrant.QUADRANT_IV -> (31..38).toList()
}

fun ToothQuadrant.isReverse() = this == ToothQuadrant.QUADRANT_I || this == ToothQuadrant.QUADRANT_III

fun Int.toToothQuadrant(): ToothQuadrant = when (this) {
    1 -> ToothQuadrant.QUADRANT_I
    2 -> ToothQuadrant.QUADRANT_II
    3 -> ToothQuadrant.QUADRANT_III
    4 -> ToothQuadrant.QUADRANT_IV
    else -> ToothQuadrant.QUADRANT_I
}
